docs/hugo/content/design/ADR-2024-10-OneOf-Resources/_index.md

```diff
@@ -0,0 +1,459 @@
+
+
+## Option 3: Change the rules for all one-of types
+
+At the moment, the only permitted properties at the root level of a one-of are the mutually exclusive properties that represent the available options. We could loosen this rule to permit other properties, those in common to all leaf types, to be present at the root level.
+
+``` mermaid
+classDiagram
+
+class Role {
+    AzureName string
+    StartDate date
+    FinishDate optional<date>
+    Student StudentProperties
+    Tutor TutorProperties
+    Teacher TeacherProperties
+    Marker MarkerProperties
+}
+
+class StudentProperties {
+    Class string
+    Grade string
+}
+
+
+class TutorProperties {
+    Class string
+    Timeslot string
+}
+
+class TeacherProperties {
+    Subject string
+    Tenure bool
+}
+
+class MarkerProperties {
+    Exam string
+    Strictness string
+}
+
+Role *-- StudentProperties : Student
+Role *-- TutorProperties : Tutor
+Role *-- TeacherProperties : Teacher
+Role *-- MarkerProperties : Marker
+```
+
+
+### Pros 
+
+* Conceptually simpler
+
+### Cons
+
+* Changing already complex one-of handling to make it more complex.
+* Requires changing the way we serialize/deserialize one-of types in non-trivial ways.
+* Potentially large blast-radius as these changes would impact on all one-of types we've already generated and released. We'd either need to take a breaking change on all these resources, or maintain two different sets of rules for one-of types.
+
+## Option 4: Special case root one-of types
+
+Preserve the existing rules for one-of types, but special case the root level of a resource spec to permit additional properties (including both `AzureName` and `OperatorSpec`) be specified alongside the one-of properties.
+
+``` mermaid
+classDiagram
+
+class Role {
+    AzureName string
+    OperatorSpec RoleOperatorSpec
+    Student StudentProperties
+    Tutor TutorProperties
+    Teacher TeacherProperties
+    Marker MarkerProperties
+}
+
+class StudentProperties {
+    Class string
+    Grade string
+    StartDate date
+    FinishDate optional<date>
+}
+
+
+class TutorProperties {
+    Class string
+    Timeslot string
+    StartDate date
+    FinishDate optional<date>
+}
+
+class TeacherProperties {
+    Subject string
+    Tenure bool
+    StartDate date
+    FinishDate optional<date>
+}
+
+class MarkerProperties {
+    Exam string
+    Strictness string
+    StartDate date
+    FinishDate optional<date>
+}
+
+Role *-- StudentProperties : Student
+Role *-- TutorProperties : Tutor
+Role *-- TeacherProperties : Teacher
+Role *-- MarkerProperties : Marker
+```
+
+
+### Pros
+
+* Limits the scope of impact of the change.
+
+### Cons
+
+* Need to change the generation of our JSON marshalling code to handle this special case.
+* Special casing is always a bit of a code smell.
+
```

Comment:
I think the other con here is that it's now less clear which properties are the "one-of" properties and which are not.

So when presented with an arbitrary resource spec like:
```
spec:
  foo:
  bar:
  baz:
  azureName:
  operatorSpec:
```

You just sort of need to know that foo/bar/baz, only one can be specified, but `azureName` and `operatorSpec` can be alongside them.

I suspect this is something that will trip people up once or twice and then they'll get the hang of it for a given resource, so it's not a huge problem, but it's really the biggest issue I see w/ this approach.